src/WebJobs.Extensions.DurableTask/Listener/DurableTaskTargetScaler.cs

```diff
@@ -0,0 +1,79 @@
+        {
+            this.functionId = functionId;
+            this.metricsProvider = metricsProvider;
+            this.scaleResult = new TargetScalerResult();
+            this.TargetScalerDescriptor = new TargetScalerDescriptor(this.functionId);
+            this.durabilityProvider = durabilityProvider;
+            this.logger = logger;
+        }
+
+        public TargetScalerDescriptor TargetScalerDescriptor { get; private set; }
+
+        private int MaxConcurrentActivities => this.durabilityProvider.MaxConcurrentTaskActivityWorkItems;
+
+        private int MaxConcurrentOrchestrators => this.durabilityProvider.MaxConcurrentTaskOrchestrationWorkItems;
+
+        public async Task<TargetScalerResult> GetScaleResultAsync(TargetScalerContext context)
+        {
+            var metrics = await this.metricsProvider.GetMetricsAsync();
+
+            // compute activityWorkers: the number of workers we need to process all activity messages
+            var workItemQueueLength = metrics.WorkItemQueueLength;
+            double activityWorkers = Math.Ceiling(workItemQueueLength / (double)this.MaxConcurrentActivities);
+
+            var serializedControlQueueLengths = metrics.ControlQueueLengths;
+            var controlQueueLengths = JsonConvert.DeserializeObject<IReadOnlyList<int>>(serializedControlQueueLengths);
+
+            var controlQueueMessages = controlQueueLengths.Sum();
+            var activeControlQueues = controlQueueLengths.Count(x => x > 0);
+
+            // compute orchestratorWorkers: the number of workers we need to process all orchestrator messages.
+            // We bound this result to be no larger than the partition count
+            var upperBoundControlWorkers = Math.Ceiling(controlQueueMessages / (double)this.MaxConcurrentOrchestrators);
+            var orchestratorWorkers = Math.Min(activeControlQueues, upperBoundControlWorkers);
+
+            int numWorkersToRequest = (int)Math.Max(activityWorkers, orchestratorWorkers);
+            this.scaleResult.TargetWorkerCount = numWorkersToRequest;
+
+            // When running on ScaleController V3, ILogger logs are forwarded to Kusto.
+            var scaleControllerLog = $"Target worker count for {this.functionId}: {numWorkersToRequest}. " +
+                $"Metrics used: workItemQueueLength={workItemQueueLength}. controlQueueLengths={serializedControlQueueLengths}. " +
+                $"maxConcurrentOrchestrators={this.MaxConcurrentOrchestrators}. maxConcurrentActivities={this.MaxConcurrentActivities}";
+
+            // target worker count should never be negative
+            if (numWorkersToRequest < 0)
+            {
+                scaleControllerLog = "Tried to request a negative worker count." + scaleControllerLog;
+                this.logger.LogError(scaleControllerLog);
+                throw new Exception(scaleControllerLog);
```

Comment:
Is throwing an exception the behavior we want? Or should we fail more gracefully?

Reply:
I can't immediately think of a more graceful solution here. My hunch is that if the scaling logic is requesting a negative number, then something is terribly wrong and we should fail exceptionally so the ScaleController can take over and decide how to deal with it. Do you have any suggestions?

Reply:
So you're saying this code is executed by the scale controller and not in the Functions host? Are there some doc comments we can add to make this clear? I think that also helps clarify my question about the direct use of ILogger.

Reply:
My understanding is that ScaleController V3 calls this directly, although @bachuv just pointed me to a [comment](https://github.com/Azure/azure-webjobs-sdk/issues/2909) that suggests that this code may also be accessed in the runtime driven scale case **as well,** which would go through the Host. So perhaps there's two cases: one mediated by the Functions Host, and one where it's accessed directly by the Scale Controller.

I'll look to get a conclusive answer on this. For now, I've pushed a series of comments reflecting my current understanding. I won't merge until we have confirmation from the ScaleController team.

Reply:
I spoke with Alexey today to get clarity on this.

The caller of this code depends on whether the user is running with or without a VNET.

If running **without** a VNET, the ScaleController process calls this directly.
If running with a VNET, then the customer needs to enable **runtime-driven scaling** and in that case it is the Host which provides calls this code for us.

I think this may complicate things a little bit, especially around throwing exceptions. We should confirm what's the preferred error handling behavior in the VNET case.

Reply:
@alrod: if runtime-driven scaling is enabled, and the DF Extension encounters an error in calculating it's target worker count (for a example, if it wrongly determines that it needs a negative target worker count), is safe to throw an exception for the Host to catch or should we handle the failure more gracefully?

Reply:
Is the code even make since to exists?

`int numWorkersToRequest = (int)Math.Max(activityWorkers, orchestratorWorkers);`
Is possible activityWorkers and orchestratorWorkers are < 0 ?